File: python/161-170/Maximum Gap.py
import math

class Solution:
    # @param num, a list of integer
    # @return an integer
    def maximumGap(self, num):
        if len(num) < 2 or min(num) == max(num):
            return 0
        a, b = min(num), max(num)
        gap = int(math.ceil((b-a)*1.0/(len(num)-1)))
        bucket = [[None, None] for _ in range((b-a)/gap+1)]
        for n in num:
            b = bucket[int((n-a)/gap)]
            b[0] = n if b[0] is None else min(b[0], n)
            b[1] = n if b[1] is None else max(b[1], n)
        bucket = [b for b in bucket if b[0] is not None]
        return max(bucket[i][0]-bucket[i-1][1] for i in range(1, len(bucket)))

items = [1,1,1,1,1,5,5,5,5,5]
s = Solution()
print s.maximumGap(items)

File: cpp/161-170/Maximum Gap.cpp
class Solution {
public:
    int maximumGap(vector<int> &num) {
        if(num.size() < 2) return 0;
        
        int minium = *min_element(num.begin(), num.end());
        int maxium = *max_element(num.begin(), num.end());
        
        int gap = (int)ceil((maxium - minium)*1.0/(num.size()-1));
        int* maxBucket = new int[num.size()];
        int* minBucket = new int[num.size()];
        
        fill(maxBucket, maxBucket+num.size(), INT_MIN);
        fill(minBucket, minBucket+num.size(), INT_MAX);
        
        for(int i = 0;i < num.size();i++)
        {
            int idx = (num[i] - minium)/gap;
            maxBucket[idx] = max(maxBucket[idx], num[i]);
            minBucket[idx] = min(minBucket[idx], num[i]);
        }
        
        int MaxGap = -1;
        int temp = minium;
        for(int i = 0;i < num.size();i++)
        {
            if(maxBucket[i] == INT_MIN)
                continue;
            MaxGap = max(MaxGap, minBucket[i] - temp);
            temp = maxBucket[i];
        }
        return MaxGap;
    }
};